### PR TITLE
Entity Resolution organization local steps

### PR DESCRIPTION
**Pull request by limetreeestate**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -23,7 +23,7 @@ add_executable(JasmineGraph main.cpp src/server/JasmineGraphServer.cpp src/serve
...
+        src/ml/trainer/JasmineGraphTrainingSchedular.cpp src/ml/trainer/JasmineGraphTrainingSchedular.h src/ml/trainer/JasminGraphTrainingInitiator.cpp src/ml/trainer/JasminGraphTrainingInitiator.h src/query/algorithms/linkprediction/JasminGraphLinkPredictor.h src/query/algorithms/linkprediction/JasminGraphLinkPredictor.cpp src/centralstore/JasmineGraphHashMapDuplicateCentralStore.cpp src/centralstore/JasmineGraphHashMapDuplicateCentralStore.h src/exception/JasmineGraphException.cpp src/exception/JasmineGraphException.h src/query/algorithms/entityresolution/BloomFilter.hpp src/query/algorithms/entityresolution/MurmurHash3.h src/query/algorithms/entityresolution/MurmurHash3.cpp src/query/algorithms/entityresolution/Kmeans.hpp src/query/algorithms/entityresolution/MinHash.hpp src/query/algorithms/entityresolution/EntityResolver.cpp src/query/algorithms/entityresolution/EntityResolver.hpp src/query/algorithms/entityresolution/Kmeans.cpp src/query/algorithms/entityresolution/MinHash.cpp)
```
Comment (miyurud) [inline CMakeLists.txt:26]: Can we divide this line into two? It is currently very lengthy.

File: CMakeLists.txt [hunk 2]
```diff
@@ -32,12 +32,16 @@ if(CMAKE_ENABLE_DEBUG)
...
+target_link_libraries(JasmineGraph /usr/local/lib/libmetis.so)
+target_link_libraries(JasmineGraph /usr/local/lib/libxerces-c-3.2.so)
 #TODO: Need to install the library separately
 target_link_libraries(JasmineGraph sqlite3)
 target_link_libraries(JasmineGraph pthread)
 target_link_libraries(JasmineGraph rdkafka)
+target_link_libraries(JasmineGraph armadillo)
+target_link_libraries(JasmineGraph stdc++)
+target_link_libraries(JasmineGraph m)
+#target_link_libraries(JasmineGraph O2)
```
Comment (miyurud) [inline CMakeLists.txt:44]: We should remove these commented lines from the PR

File: CMakeLists.txt [hunk 3]
```diff
@@ -47,12 +51,12 @@ include_directories("$ENV{HOME}/software/spdlog/include")
 include_directories("$ENV{HOME}/software/jsoncpp/include")
 include_directories("$ENV{HOME}/software/cppkafka/include")
 
-target_link_libraries(JasmineGraph /usr/local/lib/libmetis.a)
-target_link_libraries(JasmineGraph $ENV{HOME}/software/xerces-c-3.2.2/lib/libxerces-c.so)
+#target_link_libraries(JasmineGraph /usr/local/lib/libmetis.a)
```
Comment (miyurud) [inline CMakeLists.txt:54]: We should remove these commented lines from the PR

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 4]
```diff
@@ -66,7 +70,8 @@ void *frontendservicesesion(std::string masterIP, int connFd, SQLiteDBInterface
 
         if (line.compare(EXIT) == 0) {
             break;
-        } else if (line.compare(LIST) == 0) {
+        }
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:73]: We should not divide `} else if (` into two lines.

Comment (miyurud) [reply src/frontend/JasmineGraphFrontEnd.cpp:73]: This seems to be some automatic formatting issue.

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 5]
```diff
@@ -933,8 +951,126 @@ void *frontendservicesesion(std::string masterIP, int connFd, SQLiteDBInterface
...
+                graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+
+                vector<string> strParams = utils.split(designatedWorkerDetails, '|');
+                string designatedHost = strParams[0];
+                string designatedPort = strParams[1];
+
+                frontend_logger.log("Graph ID received: " + graphID, "info");
+
+                if (JasmineGraphFrontEnd::graphExistsByID(graphID, sqlite)) {
+                    frontend_logger.log("Graph with ID " + graphID + " is being deleted now", "info");
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:992]: Why do we say this particular graph is getting deleted? Entity resolution does not have any step related to graph data getting deleted.

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 6]
```diff
@@ -933,8 +951,126 @@ void *frontendservicesesion(std::string masterIP, int connFd, SQLiteDBInterface
...
+                        continue;
+                    }
+                    result_wr = write(connFd, "\n", 2);
+                    if(result_wr < 0) {
+                        frontend_logger.log("Error writing to socket", "error");
+                        loop = true;
+                        continue;
+                    }
+                } else {
+                    frontend_logger.log("Graph does not exist or cannot be deleted with the current hosts setting",
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1007]: Why do we say the graph cannot be deleted?

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 7]
```diff
@@ -933,8 +951,126 @@ void *frontendservicesesion(std::string masterIP, int connFd, SQLiteDBInterface
...
+                    }
+                    result_wr = write(connFd, "\n", 2);
+                    if(result_wr < 0) {
+                        frontend_logger.log("Error writing to socket", "error");
+                        loop = true;
+                        continue;
+                    }
+                }
+            }
+            else {
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1023]: Formatting issue with this else statement

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 8]
```diff
@@ -933,8 +951,126 @@ void *frontendservicesesion(std::string masterIP, int connFd, SQLiteDBInterface
...
+                graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+
+                vector<string> strParams = utils.split(designatedWorkerDetails, '|');
+                string designatedHost = strParams[0];
+                string designatedPort = strParams[1];
+
+                frontend_logger.log("Graph ID received: " + graphID, "info");
+
+                if (JasmineGraphFrontEnd::graphExistsByID(graphID, sqlite)) {
+                    frontend_logger.log("Graph with ID " + graphID + " is being deleted now", "info");
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1058]: Why do we say the graph is getting deleted?

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 9]
```diff
@@ -2562,3 +2698,57 @@ map<long, long> JasmineGraphFrontEnd::getOutDegreeDistributionHashMap(map<long,
     }
     return distributionHashMap;
 }
+
+void JasmineGraphFrontEnd::initiateEntityResolutionCoordinator(std::string graphID, SQLiteDBInterface sqlite, std::string masterIP, string designatedWorkerHost, string designatedWorkerPort) {
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:2702]: This function signature is too lengthy. Lets break this into two.

File: src/frontend/JasmineGraphFrontEnd.h [hunk 10]
```diff
@@ -47,6 +48,12 @@ class JasmineGraphFrontEnd {
 
     int run();
 
+    void setServer(JasmineGraphServer s);
+
+    static void initiateEntityResolutionOrg(std::string graphID, SQLiteDBInterface sqlite, std::string masterIP, std::string designatedWorkerHost, std::string designatedWorkerPort);
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.h:53]: This line is too lengthy. Lets break into two.

File: src/query/algorithms/entityresolution/BloomFilter.hpp [hunk 11]
```diff
@@ -0,0 +1,66 @@
+//
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/BloomFilter.hpp:1]: Need to have proper license header

Comment (miyurud) [inline src/query/algorithms/entityresolution/BloomFilter.hpp:1]: Need to move this file from `.hpp` to `.h` file. We had discussed about this multiple times before during offline discussions.

File: src/query/algorithms/entityresolution/BloomFilter.hpp [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        if (len < 2) {
+            char key = str[0];
+            add(&key, 1);
+        } else {
+            for (int i = 0; i < len - 2; i++) {
+                const char *key = str.substr(i, i+2).c_str();
+                add(key, 2);
+            }
+        }
+//
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/BloomFilter.hpp:31]: Remove commented lines

File: src/query/algorithms/entityresolution/BloomFilter.hpp [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+//            add(&s, 1);
+//        }
+    }
+
+    void add(const char *data, std::size_t len) {
+        auto hashValues = hash(data, len);
+
+        for (int n = 0; n < m_numHashes; n++) {
+            uint64_t pos = nthHash(n, hashValues[0], hashValues[1], m_bits.size());
+//            std::cout << pos << std::endl;
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/BloomFilter.hpp:42]: Remove the commented line

File: src/query/algorithms/entityresolution/EntityResolver.cpp [hunk 14]
```diff
@@ -0,0 +1,493 @@
+//
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/EntityResolver.cpp:1]: Add license header

File: src/query/algorithms/entityresolution/Kmeans.cpp [hunk 15]
```diff
@@ -0,0 +1,50 @@
+//
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/Kmeans.cpp:1]: Need copyright header here

File: src/query/algorithms/entityresolution/Kmeans.cpp [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+// Created by root on 4/16/21.
+//
+
+#include "Kmeans.hpp"
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/Kmeans.cpp:5]: We should change the extension to `.h`

File: src/query/algorithms/entityresolution/Kmeans.hpp [hunk 17]
```diff
@@ -0,0 +1,30 @@
+//
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/Kmeans.hpp:1]: Missing copyright header

File: src/query/algorithms/entityresolution/MinHash.cpp [hunk 18]
```diff
@@ -0,0 +1,58 @@
+//
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/MinHash.cpp:1]: Missing copyright header

File: src/query/algorithms/entityresolution/MinHash.cpp [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+
+#include "MinHash.hpp"
+
+arma::Col<short> MinHash::generateCRV(arma::Mat<float> &data, uint8_t d, bool quietPrint) {
+    display("Creating CRV", quietPrint);
+    //Calculate density vector
+    display("Create density vector", quietPrint);
+    arma::Col<float> denVec = getDensity(data);
+//        denVec.print();
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/MinHash.cpp:12]: Lets remove the commented unused code segments

File: src/query/algorithms/entityresolution/MinHash.hpp [hunk 20]
```diff
@@ -0,0 +1,52 @@
+//
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/MinHash.hpp:1]: Missing copyright header

File: src/query/algorithms/entityresolution/MurmurHash3.h [hunk 21]
```diff
@@ -0,0 +1,35 @@
+//
```
Comment (miyurud) [inline src/query/algorithms/entityresolution/MurmurHash3.h:1]: Missing copyright header

File: src/server/JasmineGraphInstance.cpp [hunk 22]
```diff
@@ -231,7 +231,7 @@ bool JasmineGraphInstance::sendFileThroughService(std::string host, int dataPort
         std::cerr << "ERROR connecting to port " << dataPort << std::endl;
     }
 
-    fileName =  "jasminegraph-local_trained_model_store/"+fileName;
+//    fileName =  "jasminegraph-local_trained_model_store/"+fileName;
```
Comment (miyurud) [inline src/server/JasmineGraphInstance.cpp:234]: Remove the commented line

File: src/server/JasmineGraphInstanceService.cpp [hunk 23]
```diff
@@ -1378,18 +1390,447 @@ void *instanceservicesession(void *dummyPt) {
...
+            instance_logger.log("Received Graph ID: " + graphID, "info");
+            write(connFd, JasmineGraphInstanceProtocol::SEND_PARTITION_ID.c_str(),
+                  JasmineGraphInstanceProtocol::SEND_PARTITION_ID.size());
+            instance_logger.log("Sent : " + JasmineGraphInstanceProtocol::SEND_PARTITION_ID, "info");
+            bzero(data, INSTANCE_DATA_LENGTH);
+            read(connFd, data, INSTANCE_DATA_LENGTH);
+            string partitionID = (data);
+            instance_logger.log("Received partition ID: " + partitionID, "info");
+
+            //pthread_mutex_lock(&file_lock);
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:1424]: We can remove the commented lines isn't it?

File: src/server/JasmineGraphInstanceService.cpp [hunk 24]
```diff
@@ -2181,4 +2654,587 @@ void JasmineGraphInstanceService::trainPartition(string trainData){
...
+        //Change pointer to next paty in the chain
+        nextParty = pairwiseCommonEntities[currentParty].begin()->first;
+    }
+
+    return partyCommonEntityMap;
+}
+
+void JasmineGraphInstanceService::entityRes(string trainData) {
+
+//    Utils utils;
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:3041]: We should either uncomment and use these commented code or permanently delete them.

File: src/server/JasmineGraphServer.cpp [hunk 25]
```diff
@@ -2360,6 +2362,1019 @@ int JasmineGraphServer::removePartitionThroughService(string host, int port, str
...
+                }
+
+            }
+
+
+        }
+    }
+}
+
+int JasmineGraphServer::signalOrganizations(string organizationHost, string designatedWorkerHost, string designatedWorkerPort, string graphID, string masterIP) {
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2559]: Too lengthy function declaration. Lets break into two.

File: src/server/JasmineGraphServer.cpp [hunk 26]
```diff
@@ -2360,6 +2362,1019 @@ int JasmineGraphServer::removePartitionThroughService(string host, int port, str
...
+    Utils utils;
+    Logger instance_logger;
+    bool result = true;
+    std::cout << pthread_self() << " host : " << organizationHost << "port : " << 7777 << std::endl;
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2563]: Shall we remove this cout statement? I do not see any use of it.

File: src/server/JasmineGraphServer.cpp [hunk 27]
```diff
@@ -2360,6 +2362,1019 @@ int JasmineGraphServer::removePartitionThroughService(string host, int port, str
...
+            server_logger.log("Sent : Graph ID " + graphID, "info");
+
+            bzero(data, 301);
+            read(sockfd, data, 300);
+            response = (data);
+            response = utils.trim_copy(response, " \f\n\r\t\v");
+
+            if (response.compare(JasmineGraphInstanceProtocol::SEND_PARTITION_ID) == 0) {
+                server_logger.log("Received : " + JasmineGraphInstanceProtocol::SEND_PARTITION_ID, "info");
+//                result_wr = write(sockfd, clusters.c_str(), clusters.size());
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:3224]: Remove the commented lines

File: src/server/JasmineGraphServer.cpp [hunk 28]
```diff
@@ -2360,6 +2362,1019 @@ int JasmineGraphServer::removePartitionThroughService(string host, int port, str
...
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:3351]: Remove the commented lines

**Review by miyurud — CHANGES_REQUESTED** [review]
I added few more comments. Please address them.

File: CMakeLists.txt (hunk 1 above)
Comment (limetreeestate) [reply CMakeLists.txt:26]: Line divided into shorter lines

File: CMakeLists.txt (hunk 3 above)
Comment (limetreeestate) [reply CMakeLists.txt:54]: Commented line removed

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 5 above)
Comment (limetreeestate) [reply src/frontend/JasmineGraphFrontEnd.cpp:992]: Neglected to change logging message when copied from another code segment. Changed to log appropriate message

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 6 above)
Comment (limetreeestate) [reply src/frontend/JasmineGraphFrontEnd.cpp:1007]: Changed to log appropriate message

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 8 above)
Comment (limetreeestate) [reply src/frontend/JasmineGraphFrontEnd.cpp:1058]: Changed to log appropriate message

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 9 above)
Comment (limetreeestate) [reply src/frontend/JasmineGraphFrontEnd.cpp:2702]: Function declaration split and adjusted

File: src/frontend/JasmineGraphFrontEnd.h (hunk 10 above)
Comment (limetreeestate) [reply src/frontend/JasmineGraphFrontEnd.h:53]: Function declaration split and length adjusted

File: src/query/algorithms/entityresolution/BloomFilter.hpp (hunk 11 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/BloomFilter.hpp:1]: JasmineGraph License header added with year as 2021

Comment (limetreeestate) [reply src/query/algorithms/entityresolution/BloomFilter.hpp:1]: Changed to .h extension

File: src/query/algorithms/entityresolution/BloomFilter.hpp (hunk 12 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/BloomFilter.hpp:31]: Commented lines removed

File: src/query/algorithms/entityresolution/BloomFilter.hpp (hunk 13 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/BloomFilter.hpp:42]: Lines removed

File: src/query/algorithms/entityresolution/EntityResolver.cpp (hunk 14 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/EntityResolver.cpp:1]: JasmineGraph License header added with year as 2021

File: src/query/algorithms/entityresolution/Kmeans.cpp (hunk 15 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/Kmeans.cpp:1]: JasmineGraph License header added with year as 2021

File: src/query/algorithms/entityresolution/Kmeans.cpp (hunk 16 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/Kmeans.cpp:5]: Extension changed to .h

File: src/query/algorithms/entityresolution/Kmeans.hpp (hunk 17 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/Kmeans.hpp:1]: JasmineGraph License header added with year as 2021

File: src/query/algorithms/entityresolution/MinHash.cpp (hunk 18 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/MinHash.cpp:1]: JasmineGraph License header added with year as 2021

File: src/query/algorithms/entityresolution/MinHash.cpp (hunk 19 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/MinHash.cpp:12]: Commented code removed

File: src/query/algorithms/entityresolution/MinHash.hpp (hunk 20 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/MinHash.hpp:1]: JasmineGraph License header added with year as 2021

File: src/query/algorithms/entityresolution/MurmurHash3.h (hunk 21 above)
Comment (limetreeestate) [reply src/query/algorithms/entityresolution/MurmurHash3.h:1]: Author's copyright header added

File: src/server/JasmineGraphInstanceService.cpp (hunk 23 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:1424]: This code is not a part of my changes, should I make changes here?

File: src/server/JasmineGraphServer.cpp (hunk 25 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:2559]: Declaration broken and length adjusted

File: src/server/JasmineGraphServer.cpp (hunk 26 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:2563]: It  is used for server logging purposes in every scenario where a socket connection is created

**Comment by miyurud** [comment]
Closing this PR as the same functionality is handled by https://github.com/miyurud/jasminegraph/pull/136

**Pull request closed without merging**